be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp

```diff
@@ -0,0 +1,228 @@
+    };
+    RETURN_IF_ERROR(set_call_back_function(state, io_executor));
+
+    return Status::OK();
+}
+
+Status SpillableHashJoinBuildOperator::publish_runtime_filters(RuntimeState* state) {
+    // publish empty runtime filters
+
+    // Building RuntimeBloomFilter need to know the initial hash table size and all join keys datas.
+    // It usually involves re-reading all the data that has been spilled
+    // which cannot be streamed process in the spill scenario when build phase is finished
+    // (unless FE can give an estimate of the hash table size), so we currently empty all the hash tables first
+
+    size_t merger_index = _driver_sequence;
+    // make sure ht_row_count exceed max runtime filters
+    auto ht_row_count = _partial_rf_merger->limit() + 1;
+    RuntimeInFilters partial_in_filters;
+    RuntimeBloomFilters partial_bloom_filters;
+    auto& partial_bloom_filter_build_params = _join_builder->get_runtime_bloom_filter_build_params();
+
+    ASSIGN_OR_RETURN(auto merged,
+                     _partial_rf_merger->add_partial_filters(merger_index, ht_row_count, std::move(partial_in_filters),
+                                                             std::move(partial_bloom_filter_build_params),
+                                                             std::move(partial_bloom_filters)));
+    if (merged) {
+        RuntimeInFilterList in_filters;
+        RuntimeBloomFilterList bloom_filters;
+
+        // publish runtime bloom-filters
+        state->runtime_filter_port()->publish_runtime_filters(bloom_filters);
+        // move runtime filters into RuntimeFilterHub.
+        runtime_filter_hub()->set_collector(_plan_node_id, std::make_unique<RuntimeFilterCollector>(
+                                                                   std::move(in_filters), std::move(bloom_filters)));
+    }
+    return Status::OK();
+}
+
+Status SpillableHashJoinBuildOperator::append_hash_columns(const ChunkPtr& chunk) {
+    auto factory = down_cast<SpillableHashJoinBuildOperatorFactory*>(_factory);
+    const auto& build_partition = factory->build_side_partition();
+
+    size_t num_rows = chunk->num_rows();
+    auto hash_column = spill::SpillHashColumn::create(num_rows);
+    auto& hash_values = hash_column->get_data();
+
+    // TODO: use different hash method
+    for (auto& expr_ctx : build_partition) {
+        ASSIGN_OR_RETURN(auto res, expr_ctx->evaluate(chunk.get()));
+        res->fnv_hash(hash_values.data(), 0, num_rows);
+    }
+    chunk->append_column(std::move(hash_column), -1);
+    return Status::OK();
+}
+
+bool SpillableHashJoinBuildOperator::is_finished() const {
+    return _is_finished;
+}
+
+Status SpillableHashJoinBuildOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
+    DeferOp update_revocable_bytes{
+            [this]() { set_revocable_mem_bytes(_join_builder->hash_join_builder()->hash_table().mem_usage()); }};
```

Comment:
should below Line:137 which is in NO_SPILL?

Reply:
set_revocable_mem_bytes means how much memory we have available to be freed